autograder/handgrading/tests/test_views/test_handgrading_result_views.py

```diff
@@ -615,6 +616,91 @@ def test_student_delete_result_permission_denied(self) -> None:
             self.handgrading_result, self.client, self.student, self.url)
 
 
+@dataclass
+class _GroupHGDataConfig:
+    submission_status: ag_models.Submission.GradingStatus =\
+        ag_models.Submission.GradingStatus.finished_grading
+    has_handgrading_result: bool = True
+
+    def __post_init__(self):
+        if self.has_handgrading_result:
+            # If the submission has a handgrading result, it must have a finished_grading status
+            assert self.submission_status == ag_models.Submission.GradingStatus.finished_grading, \
+                'Submissions with a handgrading result must have finished_grading status'
+
+        if self.submission_status != ag_models.Submission.GradingStatus.finished_grading:
+            # If the submission does not have a finished_grading status, it must not
+            # have a handgrading result
+            assert not self.has_handgrading_result, \
+                ('Submissions with a status other than'
+                 ' "finished_grading" must not have a handgrading result')
+
+    def create_submission(self, group: ag_models.Group) -> ag_models.Submission | None:
+        """
+        Create a submission based on the configuration.
+        If submission_status is None, return None.
+        """
+        if self.submission_status is None:
+            return None
+        return obj_build.make_submission(group=group, status=self.submission_status)
+
+    @staticmethod
+    def format_group_data(group, handgrading_result: dict | None,
+                          has_handgradeable_submission: bool) -> dict:
+        """Format the group data for comparison with the response data."""
+        data = group.to_dict()
+        data['handgrading_result'] = handgrading_result
+        data['member_names'].sort()
+        data['has_handgradeable_submission'] = has_handgradeable_submission
+        return data
+
+    def generate_expected_data(self, group: ag_models.Group,
+                               rubric: hg_models.HandgradingRubric) -> dict:
+        """Generate the expected data base on submission status and handgrading result status."""
+        submission = self.create_submission(group)
+        match self.has_handgrading_result, self.submission_status:
```

Comment:
I agree that using match makes sense here. To make it more explicit, I would match on the whole object, which then lets you use keyword matching in the cases. Something like:

```python3
match self:
    case _GroupHGDataConfig(has_handgrading_result=True, self.submission_status=ag_models.Submission.GradingStatus.finished_grading):
       ...
    case _GroupHGDataConfig(has_handgrading_result=False, self.submission_status=ag_models.Submission.GradingStatus.finished_grading):
        ...
    case _GroupHGDataConfig(has_handgrading_result=False):
        ...
    case _:
        assert False, 'Unexpected data configuration'
```